### ISSUE TITLE
Regarding Kotlin Contracts

### ISSUE THREAD

**Issue by nlbuescher** [edvin/tornadofx#939]
So currently members instantiated from the builders (eg `vbox`, `stackpane`, `textfield`, etc) cannot be declared as `val`, which means having to declare them as `var ... by singleAssign()` like so:

```kotlin
class TestView : View() {
    var hello: Text by singleAssign()
    override val root = stackpane {
        hello = text("Hello, World!")
    }
}
```

The reason that we cannot change `hello` to `val hello: Text` here is that the compiler doesn't know that the lambda being passed to `stackpane(...)` is being called exactly once and gives a compile error about possible reassignment of `val`.

With contracts introduced in Kotlin 1.3, we could theoretically modify the function `stackpane` like so:

```kotlin
fun EventTarget.stackpane(initialChildren: Iterable<Node>? = null, op: Group.() -> Unit = {}): StackPane {
    contract {
        callsInPlace(op, InvocationKind.EXACTLY_ONCE)
    }
    return opcr(this, StackPane().apply { if (initialChildren != null) children.addAll(initialChildren) }, op)
}
```

This would allow setting final members from the builders, eliminating the need for `by singleAssign()`.

**Comment by edvin** [comment]
Almost every time I see someone use `singleAssign` they are doing something wrong or at least far from best practice. The most common one is to save a reference to an UI control to be able to manipulate it later, but what they really should have been manipulating is a property bound to the UI control. My first though is that I'm not keen on facilitating this anti pattern.

`InvocationKind.EXACTLY_ONCE` sounds like it will only allow the `stackpane` builder to be called once, which would restrict the UI to a single stackpane at any level, is that right? If that's the case, this is also a no go for that reason :)

**Comment by nlbuescher** [comment]
1. Makes sense. My thought was that since `singleAssign` exists as a workaround for not being able to use final members for the reason above, this provides the ability to set final members without a workaround, enabling the removal of `singleAssign`. I think the important points here are that `singleAssign` already exists, and contracts would enable doing the *same thing* in a more Kotlinic way.

2. That's incorrect. `InvocationKind.EXACTLY_ONCE` is not referring to the `stackpane` function, but rather to the **`op` parameter**. The contract is that whatever `op` does, it happens exactly once. For comparison:

```kotlin
// from the Kotlin stdlib
inline fun <T> T.apply(block: T.() -> Unit): T {
    contract {
        callsInPlace(block, InvocationKind.EXACTLY_ONCE)
    }
    block()
    return this
}
```

Obviously we can call `apply` however-many times we want. :)

**Comment by edvin** [comment]
That sounds interesting. I'm not familiar with this construct, sorry for my ignorance. Would this have any other side effects you can think of? We could try to implement it and run some tests.

**Comment by nlbuescher** [comment]
Not that I can think of. The feature is technically still "experimental", but seeing as it's in place in the stdlib, it's stable enough to do some tests before they actually make it stable in Kotlin 1.4 or something.

**Comment by edvin** [comment]
Good plan! I need to read up on the new features in 1.3/1.4 as well :)

**Comment by nlbuescher** [comment]
Other major features in 1.3 include stable coroutines, inline classes, unsigned types, kotlin native, and multiplatform projects. Coroutines might be applicable to this project, maybe (for async), but I don’t understand them well enough at all.

**Comment by edvin** [comment]
Thanks :) coroutines can already be used without any special support, but we'll introduce some convenience helpers for it in 2.0 for sure.

**Comment by nlbuescher** [comment]
I'll fork the project when I have some time, make the changes, run the tests and see what comes out.

**Comment by edvin** [comment]
Thanks! Very excited to hear how it goes!

**Comment by nlbuescher** [comment]
https://github.com/nlbuescher/tornadofx/tree/feature/contracts
I've forked the project and applied a contract { callsInPlace(op, EXACTLY_ONCE) } all over the place.

One error I have found: https://youtrack.jetbrains.com/issue/KT-30609

It looks like the feature is definitely not production ready yet :)

**Comment by edvin** [comment]
Thank you for putting in the time and investigating this. Let's keep track of KT-30609 and revisit later!

**Comment by nlbuescher** [comment]
Update: [KT-30609](https://youtrack.jetbrains.com/issue/KT-30609) resolved as duplicate of [KT-29385](https://youtrack.jetbrains.com/issue/KT-29385).

Currently a workaround seems to be to make the contracted function `inline`. I'll try that and see if the issue persists after packaging; `run` from the stdlib is defined similarly and does not cause this issue, so it might only be an issue only while compiling the library.

**Comment by edvin** [comment]
Added `inline` to every builder might increase the size of the compiled end user code with a lot, so that's probably not a good workaround.

**Comment by nlbuescher** [comment]
Obviously. :) I added `inline` to hbox to get the project to compile, and then several tests failed, so I'm back to starting from `master` and adding contracts til the tests fail so I can figure out what the problem is and report some more contract bugs :)

The error messages that get spit out are all along the lines of "javafx.scene.Node is not assignable to kotlin/jvm/Ref$ObjectRef", and they only happen when I run tests, not when I compile.

My main goal now that we know it fails, is to figure out how and why it fails so it can be tested against.

**Comment by SchweinchenFuntik** [comment]
I made a test branch, with a test as an example. `KotlinContractsTest`
https://github.com/SchweinchenFuntik/tornadofx/tree/kotlinContracts

**Comment by edvin** [comment]
Great! Let us know about your findings!

**Comment by SchweinchenFuntik** [comment]
need to wait to fix: https://youtrack.jetbrains.com/issue/KT-30609, or make inline functions

**Comment by SchweinchenFuntik** [comment]
I had a mistake. I did not update the plugin in the ideas, we are waiting for the release 1.3.60

**Comment by SchweinchenFuntik** [comment]
new bug: https://youtrack.jetbrains.com/issue/KT-34048
